pkg/services/ngalert/schedule/schedule.go

```diff
@@ -653,7 +635,38 @@ func (r *alertRuleRegistry) keyMap() map[models.AlertRuleKey]struct{} {
 type alertRuleInfo struct {
 	evalCh  chan *evalContext
 	stopCh  chan struct{}
-	version int64
+	mtx     sync.Mutex
+	stopped bool // indicates that the stopCh was closed and therefore the rule evaluation routine should exit
+}
+
+func newAlertRuleInfo() *alertRuleInfo {
+	return &alertRuleInfo{evalCh: make(chan *evalContext), stopCh: make(chan struct{}), stopped: false}
+}
+
+// Stop signals the rule evaluation routine to stop via stopCh and sets flag stopped to true. Does nothing If the loop is stopped
+func (a *alertRuleInfo) Stop() bool {
+	a.mtx.Lock()
+	defer a.mtx.Unlock()
+	if a.stopped {
+		return false
+	}
+	close(a.stopCh)
+	a.stopped = true
+	return true
+}
+
+// Eval signals the rule evaluation routine to perform the evaluation of the rule. Does nothing if the loop is stopped
+func (a *alertRuleInfo) Eval(t time.Time, version int64) bool {
+	a.mtx.Lock()
+	defer a.mtx.Unlock()
+	if a.stopped {
+		return false
+	}
+	a.evalCh <- &evalContext{
```

Comment:
`eval.Ch` is unbuffered and so this operation will block until `ruleRoutine` has completed it's most recent evaluation of the alert rule. That's often not an issue except we have also acquired the mutex at the start of the function?

Reply:
I think the mutex here is still useful. Yes, it will block the other "commands". The reason for having it is that it guards against the following situation
![image](https://user-images.githubusercontent.com/25988953/140766397-9f8100b6-ed26-4fa4-917c-947c4deb5f69.png)

It is possible because of the select that picks a channel that has data in a pseudo-random way (https://golang.org/ref/spec#Select_statements). I am not 100% sure whether it works the same way in the case of a closed channel, though.